Lib/asyncio/streams.py

```diff
@@ -154,27 +153,25 @@ def connection_lost(self, exc):
         # Wake up the writer if currently paused.
         if not self._paused:
             return
-        waiter = self._drain_waiter
-        if waiter is None:
-            return
-        self._drain_waiter = None
-        if waiter.done():
-            return
-        if exc is None:
-            waiter.set_result(None)
-        else:
-            waiter.set_exception(exc)
+
+        for waiter in self._drain_waiters:
+            if not waiter.done():
+                if exc is None:
+                    waiter.set_result(None)
+                else:
+                    waiter.set_exception(exc)
 
     async def _drain_helper(self):
         if self._connection_lost:
             raise ConnectionResetError('Connection lost')
         if not self._paused:
             return
-        waiter = self._drain_waiter
-        assert waiter is None or waiter.cancelled()
         waiter = self._loop.create_future()
-        self._drain_waiter = waiter
-        await waiter
+        self._drain_waiters.append(waiter)
+        try:
+            await waiter
+        finally:
+            self._drain_waiters.remove(waiter)
```

Comment:
Interesting, this might become O(N) if there are many (N) waiters. That suggests maybe we should use a set. OTOH a deque helps preserve the order in which waiters are woken up, so I'm not sure. On the third hand, maybe just use a plain list?

Reply:
I wish sets in Python preserved insertion order as dict but they don't. Everywhere `asyncio` uses deque for waiters so better left for consistency then.